### DISCUSSION TITLE
Typing state when using typegen?

### DISCUSSION THREAD

**Discussion by mimi-stripe** [statelyai/xstate#3532] [Q&A]
Hi! I'm looking into typegen and was wondering if there is a way to get explicit types for `state` just how you can define the types for Context and Events within the `schema` object. When using typegen you can't pass generics into `CreateMachine` so I haven't been able to figure out a way to get the Interpreter's `state` type to be anything other than `any` which makes it difficult to use it in any logic. Please let me know if I'm missing something here! I did see this thread (https://github.com/statelyai/xstate/discussions/3473) specifically about the `children` field, but I'm wondering in general about the `state` itself. For example, in the docs for Typestates, it looks like they are passed in as a type when using `CreateMachine` but when I try something similar I get an error about generics not being compatible with typegen: 

```Generics cannot be specified using type parameters when using typegen. Use the schema property of the machine config instead```

**Comment by Andarist** [comment]
What do you refer to as `state` here? Could you prepare a runnable repro of your problem that would showcase what you are trying to do?

**Comment by mimi-stripe** [comment]
Here's an attempt at simplifying it down!

```
import {createMachine} from 'xstate';
import {useActor, useInterpret} from '@xstate/react';

type Context = {
  myContext: string;
};

type Events = {type: 'MY_EVENT'}

const myMachine =
  createMachine(
    {
      schema: {
        context: {} as Context,
        events: {} as Events,
        services: {} as {
          myService: {
            data: {myContext: string};
          };
        },
      },
      context: {myContext: ''},
      tsTypes: {} as import('./MyMachine.typegen').Typegen0,
      id: 'myMachine',
      type: 'parallel',
      states: {
        stateOne: {
          initial: 'childOne',
          states: {
            childOne: {},
          },
        },
        stateTwo: {
          initial: 'childTwo',
          states: {
            childTwo: {
              invoke: {
                src: 'myService',
                id: 'myService',
              },
            },
          },
        },
      },
    },
    {
      services: {
        myService: () => {
          return new Promise(() => {});
        },
      },
    },
  );

// The type of this is Interpreter<Context, any, Events, ...>
const myService = useInterpret(myMachine, {
    actions: {
      myAction: () => {},
    },
  });

// The type of this is something like State<Context, Events, any, ...>
const [state] = useActor(myService);

// Error here: Property 'stateOne' does not exist on type 'StateValue'. Property 'stateOne' does not exist on type 'string'.
return <>{state.value.stateOne}</>;
```

**Reply by Andarist** [reply]
Oh, so you want to type `state.value` - here are some related PRs to this feature request: https://github.com/statelyai/xstate/pull/3342,  https://github.com/statelyai/xstate-tools/pull/160 

I wonder though - how are you using `state.value` in your application? I actually have never used it, I've always reached for `.matches`.

**Reply by mimi-stripe** [reply]
So far just playing around with the library which is why I was trying to use state.value, but I was also wondering why the State type is any for both the returned service from `useInterpret` and useActor (it looks like the types are `Interpreter<Context, any, Events, ...>` and `State<Context, Events, any, ...>`). Looks like that might be solved by the PR's you linked though? Essentially looking for parity to explicitly passing `<Context, Events, State>` to CreateMachine, since that isn't compatible with typegen, to decide if I should use typegen or not!

**Reply by Andarist** [reply]
Those `any`s correspond to `TStateSchema` generic which we no longer recommend using. To avoid breaking consumers using this generic we just provide `any` there when it's not used at all anyway. `.value`'s type probably could be computed from that generic in the past but I'm not sure if it actually ever has been done (it definitely won't be the source of this information in the future)
